Kernel/Graphics/Intel/DisplayPortAuxiliaryConnector.cpp

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2022, Liav A. <liavalb@hotmail.co.il>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Arch/x86/IO.h>
+#include <Kernel/Graphics/Intel/DisplayPortAuxiliaryConnector.h>
+#include <Kernel/PhysicalAddress.h>
+
+namespace Kernel {
+
+enum class GMBusStatus {
+    TransactionCompletion,
+    HardwareReady
+};
+
+enum DisplayPortAuxiliary {
+    Wait = 1,
+    Stop = 4,
+};
+
+ErrorOr<NonnullOwnPtr<DisplayPortAuxiliaryConnector>> DisplayPortAuxiliaryConnector::create_with_physical_address(PhysicalAddress display_port_auxiliary_start_address)
+{
+    auto registers_mapping = TRY(Memory::map_typed<volatile UnifiedDisplayPortAuxChannelRegisters>(display_port_auxiliary_start_address, sizeof(UnifiedDisplayPortAuxChannelRegisters), Memory::Region::Access::ReadWrite));
+    return adopt_nonnull_own_or_enomem(new (nothrow) DisplayPortAuxiliaryConnector(move(registers_mapping)));
+}
+
+DisplayPortAuxiliaryConnector::DisplayPortAuxiliaryConnector(Memory::TypedMapping<volatile UnifiedDisplayPortAuxChannelRegisters> all_display_port_aux_registers)
+    : m_all_display_port_aux_registers(move(all_display_port_aux_registers))
+{
+}
+
+IntelGraphics::DisplayPortAuxChannelRegisters volatile& DisplayPortAuxiliaryConnector::aux_channel_by_port_index(PortIndex port_index) const
+{
+    VERIFY(m_access_lock.is_locked());
+    switch (port_index) {
+    case PortIndex::PortA:
+        return *((IntelGraphics::DisplayPortAuxChannelRegisters volatile*)m_all_display_port_aux_registers.base_address().offset(__builtin_offsetof(UnifiedDisplayPortAuxChannelRegisters, set1)).as_ptr());
+    case PortIndex::PortB:
+        return *((IntelGraphics::DisplayPortAuxChannelRegisters volatile*)m_all_display_port_aux_registers.base_address().offset(__builtin_offsetof(UnifiedDisplayPortAuxChannelRegisters, set2)).as_ptr());
```

Comment:
This feels very strange, especially with the c-style cast. Why do we need to cast, and why do we need to use `offsetof` ?

Can we create an object that represents a view onto the "all_aux_registers" mapping for the AuxChannelRegisters that just lets us do `return m_aux_channel_registers->set2` that will automatically do a volatile read?